h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java

```diff
@@ -716,31 +716,31 @@ public Frame scoreContributions(Frame frame, Key<Frame> destination_key, Job<Fra
   @Override
   public Frame scoreContributions(Frame frame, Key<Frame> destination_key, Job<Frame> j, ContributionsOptions options, Frame backgroundFrame) {
     Log.info("Starting contributions calculation for " + this._key + "...");
-    Frame adaptedFrame = null;
-    Frame adaptedBgFrame = null;
-    try {
-      if (null == backgroundFrame)
-        return scoreContributions(frame, destination_key, j, options);
-      adaptedFrame = new Frame(frame);
-      adaptTestForTrain(adaptedFrame, true, false);
-      DKV.put(adaptedFrame);
-      adaptedBgFrame = new Frame(backgroundFrame);
-      adaptTestForTrain(adaptedBgFrame, true, false);
-      DKV.put(adaptedBgFrame);
-      
-      DataInfo di = model_info().dataInfo();
-      assert di != null;
-      final String[] featureContribNames = ContributionsOutputFormat.Compact.equals(options._outputFormat) ?
-              _output.features() : di.coefNames();
-      final String[] outputNames = ArrayUtils.append(featureContribNames, "BiasTerm");
-
-
-      return new PredictTreeSHAPWithBackgroundTask(di, model_info(), _output, options, 
-              adaptedFrame, adaptedBgFrame, options._outputPerReference, options._outputSpace)
-              .runAndGetOutput(j, destination_key, outputNames);
+    try (Scope.Safe s = Scope.safe(frame, backgroundFrame)) {
+      Frame contributions;
+      if (null == backgroundFrame) {
+        contributions = scoreContributions(frame, destination_key, j, options);
+      } else {
+        Frame adaptedFrame = new Frame(frame);
+        adaptTestForTrain(adaptedFrame, true, false);
```

Comment:
@tomasfryda, is there any reason why you use `adaptTestForTrain` here whereas you use `adaptTestForScore` in other `scoreContributions` implementations?
The frame is implicitely tracked in the later but this needs to be done explicitely with the former… also, as we're in scoring logic, I'm surprised.

Reply:
switching to `adaptFrameForScore` simplifies the code a bit, and tests are passing